ambari-common/src/main/python/resource_management/libraries/functions/mpack_manager_helper.py

```diff
@@ -65,24 +60,22 @@ def get_component_rundir_path(mpack_name, instance_name, module_name, components
   :raises ValueError if the parameters doesn't match the mpack or instances structure
   """
 
-  run_json = get_run_dir(mpack_name, instance_name, subgroup_name, module_name,
+  return get_run_dir(mpack_name, instance_name, subgroup_name, module_name,
                            {components_instance_type: [component_instance_name]})
 
-  return run_json[COMPONENTS_PLURAL_KEY_NAME][components_instance_type.lower()][COMPONENT_INSTANCES_PLURAL_KEY_NAME][
-    component_instance_name][RUN_DIR_KEY_NAME]
 
 def get_component_target_path(mpack_name, instance_name, module_name, components_instance_type,
                               subgroup_name='default', component_instance_name='default'):
   """
   :returns the single string that contains the path to the mpack component folder of given component instance
   :raises ValueError if the parameters doesn't match the mpack or instances structure
   """
-
+  dirs = set()
   instances_json = list_instances(mpack_name, instance_name, subgroup_name, module_name,
                                   {components_instance_type: [component_instance_name]})
-
-  return instances_json[COMPONENTS_PLURAL_KEY_NAME][components_instance_type.lower()][
-    COMPONENT_INSTANCES_PLURAL_KEY_NAME][component_instance_name][PATH_KEY_NAME]
+  walk_mpack_dict(instances_json, PATH_KEY_NAME, dirs)
+  return [dir for dir in dirs if
+          (mpack_name == None or mpack_name.lower() in dir) and (instance_name == None or instance_name.lower() in dir)]
 
 
 def get_versions(mpack_name, instance_name, module_name, components_instance_type,
```

Comment:
I think you also need to make changes to this method, no?

Reply:
I think so, I will also add a new UT there. Thanks.